pkg/stats/collector.go

```diff
@@ -211,154 +207,187 @@ func (s *BufferedCollector) getInstallationID() string {
 
 	// hash or clear extended fields on the event
 	if s.extended {
 		ev = ev.HashExtended()
 	} else {
 		ev = ev.ClearExtended()
 	}
-	s.writes <- Metric{Event: ev, Value: count}
+	s.updatesCh <- Metric{Event: ev, Value: count}
 }
 
 func (s *BufferedCollector) CollectEvent(ev Event) {
 	s.CollectEvents(ev, 1)
 }
 
-func (s *BufferedCollector) isCtxCancelled() bool {
-	return atomic.LoadInt32(&s.ctxCancelled) == 1
+func (s *BufferedCollector) isClosed() bool {
+	return atomic.LoadInt32(&s.closed) == 1
 }
 
 func (s *BufferedCollector) Run(ctx context.Context) {
-	if atomic.LoadInt32(&s.runCalled) != 0 {
-		return
-	}
-	atomic.StoreInt32(&s.runCalled, 1)
-	go func() {
-		for {
-			select {
-			case w := <-s.writes:
-				// collect events, and flush if needed by size
-				s.update(w)
-				if len(s.cache) >= s.flushSize {
-					s.flush()
-				}
-			case <-s.heartbeatTicker.Tick():
-				// collect heartbeat
-				s.incr(Event{
-					Class: "global",
-					Name:  "heartbeat",
-				})
-			case <-s.flushTicker.Tick():
-				// every N seconds, send the collected events
+	ctx, s.cancel = context.WithCancel(ctx)
+	const backgroundLoops = 2
+	s.wg.Add(backgroundLoops)
+	go s.updateMetricsLoop(ctx)
+	go s.sendMetricsLoop(ctx)
+}
+
+func (s *BufferedCollector) sendMetricsLoop(ctx context.Context) {
+	defer s.wg.Done()
+	s.processSendCh(ctx)
+}
+
+func (s *BufferedCollector) updateMetricsLoop(ctx context.Context) {
+	defer func() {
+		close(s.sendCh)
+		s.wg.Done()
+	}()
+	for {
+		select {
+		case w := <-s.updatesCh:
+			// collect events, and flush if needed by size
+			s.update(w)
+			if len(s.cache) >= s.flushSize {
 				s.flush()
-			case <-ctx.Done(): // we're done
-				close(s.done)
-				return
 			}
+
+		case <-s.heartbeatTicker.Tick():
+			// collect heartbeat
+			s.incr(Event{
+				Class: "global",
+				Name:  "heartbeat",
+			})
+
+		case <-s.flushTicker.Tick():
+			// every N seconds, send the collected events
+			s.flush()
+			// update runtime statistics
+			s.handleRuntimeStats()
+
+		case <-ctx.Done(): // we're done
+			return
 		}
-	}()
+	}
 }
 
+// flush send the current cache information using SendEvents
 func (s *BufferedCollector) flush() {
-	s.handleRuntimeStats()
-	if len(s.cache) == 0 {
-		// nothing to flush
+	if len(s.sendCh) > 0 {
 		return
 	}
-	metrics := makeMetrics(s.cache)
-	s.cache = make(keyIndex)
-	s.send(metrics)
+	event := s.newInputEvent()
+	if event == nil {
+		return
+	}
+	// try passing the current metrics to the sender, if we fail
+	// we keep the current cache state and retry on the next flush
+	select {
+	case s.sendCh <- event:
+		s.cache = make(keyIndex)
+	default:
+		// keep data in cache for the next time we try to send
+	}
 }
 
 func (s *BufferedCollector) Close() {
-	if atomic.LoadInt32(&s.runCalled) == 0 {
-		// nothing to do
+	// mark as closed
+	if !atomic.CompareAndSwapInt32(&s.closed, 0, 1) {
 		return
 	}
 
-	// wait for main loop to exit
-	<-s.done
+	// wait until no more in flight requests
```

Comment:
What if there's always more incoming requests?

Reply:
we set the 'stop' flag so no new request will send anything on this channel